gcloud/translate/client.py

```diff
@@ -0,0 +1,198 @@
+                  all will be present in all cases)
+
+                  * ``confidence``: The confidence in language detection, a
+                    float between 0 and 1.
+                  * ``input``: The corresponding input value.
+                  * ``language``: The detected language (as an ISO 639-1
+                    language code).
+
+                  If multiple languages are detected for a given input, then
+                  the there will be a list of dictionaries (instead of a single
+                  dictionary) for that queried value.
+        :raises: :class:`ValueError <exceptions.ValueError>` if the number of
+                 detections is not equal to the number of values.
+                 :class:`ValueError <exceptions.ValueError>` if a value
+                 produces a list of detections with 0 or multiple results
+                 in it.
+        """
+        query_params = [('key', self.key)]
+        query_params.extend(('q', _to_bytes(value, 'utf-8'))
+                            for value in values)
+        response = self.connection.api_request(
+            method='GET', path='/detect', query_params=query_params)
+        detections = response.get('data', {}).get('detections', ())
+
+        if len(values) != len(detections):
+            raise ValueError('Expected same number of values and detections',
+                             values, detections)
+
+        for index, value in enumerate(values):
+            # Empirically, even clearly ambiguous text like "no" only returns
+            # a single detection, so we replace the list of detections with
+            # the single detection contained.
+            if len(detections[index]) == 1:
+                detections[index] = detections[index][0]
+            else:
+                message = ('Expected a single detection per value, API '
+                           'returned %d') % (len(detections[index]),)
+                raise ValueError(message, value, detections[index])
+
+            detections[index]['input'] = value
+            # The ``isReliable`` field is deprecated.
+            detections[index].pop('isReliable', None)
+
+        return detections
+
+    def translate(self, *values, **kwargs):
+        """Translate a string or list of strings.
+
+        See: https://cloud.google.com/translate/v2/\
+        translating-text-with-rest
+
+        Accepted keyword arguments are:
+
+        * ``target_language`` (str): The language to translate results into.
+          This is required by the API and defaults to :data:`ENGLISH_ISO_639`.
+        * ``format`` (str): (Optional) One of ``text`` or ``html``, to specify
+          if the input text is plain text or HTML.
+        * ``source_language`` (str): (Optional) The language of the text to
+          be translated.
+        * ``customization_ids`` (list): (Optional) List of customization IDs
+          for translation. Sets the ``cid`` parameter in the query.
```

Comment:
This API pattern (`**kw` instead of named-with-defaults arguments) still smells bad:  to do it right, we should be consuming `kw` below, and raising a `TypeError` if an unknown keyword is passed.  If we dropped using `*args` for the texts to translate (requiring the user to pass them explicitly as a list) we could just make these "normal" arguments.